### PR TITLE
CATROID-48 Suspend/resume all threads on scene transition

### PR DESCRIPTION
**Pull request by rschluesselbauer**
Add scheduler state [running / paused] to pause / resume all schedulers of a scene during scene transition.

Change idToEventThreadMap to type LinkedHashMultiMap (from HashMultiMap) to ensure consistent execution order (before the order of multiple threads started at the same time was not deterministic).

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/common/ThreadScheduler.java [hunk 1]
```diff
@@ -60,7 +73,7 @@ private void startThreadsInStartQueue() {
 	private void runThreadsForOneTick(Array<Action> actions, float delta) {
 		for (int i = 0; i < actions.size; i++) {
 			Action action = actions.get(i);
-			if (action.act(delta)) {
+			if (state == RUNNING && action.act(delta)) {
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/common/ThreadScheduler.java]: I haven't looked up how unit-testable the `ThreadScheduler` is. Maybe you can test this line here with a unit test?

File: catroid/src/main/java/org/catrobat/catroid/content/Sprite.java [hunk 2]
```diff
@@ -86,7 +86,7 @@
 	public transient PenConfiguration penConfiguration = new PenConfiguration();
 	private transient boolean convertToSingleSprite = false;
 	private transient boolean convertToGroupItemSprite = false;
-	private transient Multimap<EventId, EventThread> idToEventThreadMap = HashMultimap.create();
+	private transient Multimap<EventId, EventThread> idToEventThreadMap = LinkedHashMultimap.create();
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/content/Sprite.java]: Please let me know why you switched the `HashMultimap` with a `LinkedHashMultimap` :wink:

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/SceneTransitionTest.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+import org.catrobat.catroid.uiespresso.util.rules.BaseActivityInstrumentationRule;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import static org.catrobat.catroid.uiespresso.util.UserVariableAssertions.assertUserVariableEqualsWithTimeout;
+import static org.catrobat.catroid.uiespresso.util.UserVariableAssertions.assertUserVariableNotEqualsForTimeMs;
+
+public class SceneTransitionTest {
```
Comment (robertpainsi) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/SceneTransitionTest.java]: I'll look into this in the next few days.
I think https://github.com/Catrobat/Catroid/pull/3134 could be useful to test this.

File: catroid/src/main/java/org/catrobat/catroid/content/Sprite.java (hunk 2 above)
Comment (rschluesselbauer) [reply catroid/src/main/java/org/catrobat/catroid/content/Sprite.java]: When an event is received, then all eventThreads are started in the order given by idToEventThreadMap. However the HashMultimap type didn't ensure consistent order.

This is especially problematic for testing behaviour of threads during a single tick.
If one thread should pause all other threads, but sometimes it is executed before & sometimes it is executed after the other threads, it's a pain to test.

In short it makes the execution order of the programs deterministic which is nice :-)

Comment (robertpainsi) [reply catroid/src/main/java/org/catrobat/catroid/content/Sprite.java]: Got it! Great explanation! Thanks a lot! :bowing_man:

File: catroid/src/main/java/org/catrobat/catroid/common/ThreadScheduler.java (hunk 1 above)
Comment (rschluesselbauer) [reply catroid/src/main/java/org/catrobat/catroid/common/ThreadScheduler.java]: Great idea, should be doable 👍 
I'll look into it!

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/SceneTransitionTest.java (hunk 3 above)
Comment (rschluesselbauer) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/SceneTransitionTest.java]: I added the whole test as CatrobatTestrunnerTests -if this is the prefered way I can remove this test.

File: catroid/src/main/java/org/catrobat/catroid/common/ThreadScheduler.java (hunk 1 above)
Comment (rschluesselbauer) [reply catroid/src/main/java/org/catrobat/catroid/common/ThreadScheduler.java]: Done

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/SceneTransitionTest.java (hunk 3 above)
Comment (robertpainsi) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/SceneTransitionTest.java]: Ah, stupid me! I misunderstood that PR entirely. I didn't thought that projects have to be in `.catrobat`s. D'oh!

However, I still think that the _CatrobatTestrunner_ is suited here. It verifies that, when a scene switch brick is executed, the current scene stops (and doesn't execute any further bricks) and the new scene starts.

@84n4n4, just want to make sure that the _CatrobatTestrunner_ should be used here. And mostly that the _CatrobatTestrunner_ should be used in the future for such tests.
(I'll comment on this tests later)

Comment (rschluesselbauer) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/SceneTransitionTest.java]: I checked with @84n4n4 and CatrobatTestrunner based tests are the way to go here, so I'll get rid of this espresso test.

Comment (robertpainsi) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/SceneTransitionTest.java]: :kissing_cat:

Comment (robertpainsi) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/SceneTransitionTest.java]: ~(I'll comment on this tests later)~ → LGTM!

Comment (rschluesselbauer) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/SceneTransitionTest.java]: One more thing: In my tests I assume unset variables to be equal to 0 and it works, but I suspect that could cause issues.
Should I use a scene for setting up variable values first in this case?

Comment (robertpainsi) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/SceneTransitionTest.java]: > …but I suspect that could cause issues.

Are you worrying that the default value `0` may change in the future or do you have other issues in mind?

Comment (rschluesselbauer) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/SceneTransitionTest.java]: I retested and my assumption was wrong (I thought that initial values could change based on the value of the variable during the last run). The tests are ok then 👍

Comment (robertpainsi) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/SceneTransitionTest.java]: :+1:

> so I'll get rid of this espresso test

…, then move the ticket to _Ready For Code Review and I'll approve it :tada:

Comment (robertpainsi) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/SceneTransitionTest.java]: ~_then move the ticket to Ready For Code Review_~

**Comment by wslany** [comment]
Cool, thanks a lot for this fix!
